src/Infrastructure/Event/EventStreamProcessor.php

```diff
@@ -0,0 +1,52 @@
+<?php
+
+namespace App\Infrastructure\Event;
+
+use App\Domain\Event\EventBufferInterface;
+use App\Domain\Event\EventStreamProcessorInterface;
+use Broadway\Domain\DomainEventStream;
+use Broadway\EventHandling\EventBus;
+
+class EventStreamProcessor implements EventStreamProcessorInterface
+{
+    /**
+     * @var \App\Domain\Event\EventBufferInterface
+     */
+    private $eventBuffer;
+
+    /**
+     * @var \Broadway\EventHandling\EventBus
+     */
+    private $eventBus;
+
+    /**
+     * EventStreamProcessor constructor.
+     */
+    public function __construct(
+        EventBufferInterface $eventBuffer,
+        EventBus $eventBus
+    ) {
+        $this->eventBuffer = $eventBuffer;
+        $this->eventBus = $eventBus;
+    }
+
+    public function process(DomainEventStream $stream)
+    {
+        foreach ($stream->getIterator() as $event) {
+            $this->eventBuffer->add($event);
+            if ($this->eventBuffer->isFull()) {
+                $this->publishEvents();
+            }
+        }
+
+        $this->publishEvents();
```

Comment:
this will be called twice always for no reason

Reply:
This method will be called inside the loop and after the loop.

Reply:
It will flush the buffer prematurely when it is full, that is why you see publishEvents twice.

Reply:
Or at least it is supposed to be flushed... Oops

Reply:
When it's full, then the iterator will finish and exit the loop, there's no optimisation here. Even worst you're doing a count inside a loop, something really not recommended: https://secure.php.net/manual/es/function.count.php#114924. :/